### DS_Bootcamp.Team00

#### Team TL: emmelisi
* Roman (emmelisi)
* Ariadna (parasnew)


*Data files in /data*:
1. ratings.csv: userId, movieId, rating, timestamp
2. movies.csv: movieId, title, genres
3. tags.csv: userId, movieId, tag, timestamp
4. links.csv: movieId, imdbId, tmdbId


*Project Structure*:
1. movielens_analysis: Main module with classes and functions for data analysis with tests.
2. movielens_report: Added a module for checking the main module and generating a report.
3. requirements.txt: List of required packages for the project.

In [2]:
from movielens_analysis import *

In [3]:
    R = Ratings("data/ratings.csv", "data/movies.csv")
    Rm = R.Movies(R.unity)
    Ru = R.Users(R.unity)
    M = Movies("data/movies.csv")
    T = Tags("data/tags.csv")
    L = Links('data/links.csv', 'data/movies.csv', 100)

## Анализ рейтингов. Давайте посмотрим, как распределяются оценки фильмов.

### Рейтинг годов по количеству оцениваемых фильмов
Пик пользовательских оценок пришелся на 2000 год, а наименьшее количество отслеживается в 1998 году.

In [4]:
%%timeit -r1 -n1
print(Rm.dist_by_year())

{1996: 6040, 1997: 1916, 1998: 507, 1999: 2439, 2000: 10061, 2001: 3922, 2002: 3478, 2003: 4014, 2004: 3279, 2005: 5813, 2006: 4059, 2007: 7114, 2008: 4351, 2009: 4158, 2010: 2300, 2011: 1690, 2012: 4657, 2013: 1664, 2014: 1439, 2015: 6616, 2016: 6702, 2017: 8199, 2018: 6418}
144 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Рейтинг популярности выставялемых оценок.
Самая популярная пользовательская оценка составила 4 балла, следом за ней с приличным отставанием в 6771 пункт, следует оценка 3 балла. А в самом низу рейтинга расположилась оценка 0.5 баллов с 1370 пунктами.

In [5]:
%%timeit -r1 -n1
print(Rm.dist_by_rating())

{0.5: 1370, 1.0: 2811, 1.5: 1791, 2.0: 7551, 2.5: 5550, 3.0: 20047, 3.5: 13136, 4.0: 26818, 4.5: 8551, 5.0: 13211}
27.3 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 картин по количеству оценок.
В первой тройке расположились такие легендарные картины как: Форрест Гамп, Побег из Шоушенка, Криминальное чтиво. Данным картинам удалось преодолеть отметку в 300+ оценок! 

In [6]:
%%timeit -r1 -n1 
print(Rm.top_by_num_of_ratings(10))

{'Forrest Gump': 329, 'Shawshank Redemption The': 317, 'Pulp Fiction': 307, 'Silence of the Lambs The': 279, 'Matrix The': 278, 'Star Wars: Episode IV - A New Hope': 251, 'Jurassic Park': 238, 'Braveheart': 237, 'Terminator 2: Judgment Day': 224, "Schindler's List": 220}
27.4 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 картин по их среднему рейтингу.
Здесь собраны редкие бриллианты нашей коллекции - картины, получившие исключительно 5.0 от наших пользователей. На первом месте фильм "Америка" ("Lamerica"). По информации с IMDB фильм получил 21 награду и ещё 18 номинаций. Неплохо!

In [7]:
%%timeit -r1 -n1
print(Rm.top_by_ratings(10))

{'Lamerica': 5.0, 'Heidi Fleiss: Hollywood Madam': 5.0, 'Awfully Big Adventure An': 5.0, 'Live Nude Girls': 5.0, 'In the Realm of the Senses (Ai no corrida)': 5.0, 'What Happened Was...': 5.0, 'Thin Line Between Love and Hate A': 5.0, 'Denise Calls Up': 5.0, 'Supercop 2 (Project S) (Chao ji ji hua)': 5.0, 'Entertaining Angels: The Dorothy Day Story': 5.0}
77.6 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 картин по дисперсии оценок.
Здесь мы видим 2 картины с максимальной величиной дисперсии, а именно 'Фанни и Александр' ('Fanny and Alexander') Ингмара Бергмана и 'Иваново детство' ('Ivan's Childhood') Тарковского. Те, кто знаком с этими режиссёрами, поймут почему их фильмы удостоились первого места в рейтинге самых спорных фильмов. Их фильмы можно описать словами "ничего не понятно, но очень интересно". Наш топ-10 замыкает картина с названием, которое гоорит само за себя - 'Нападение гигантских пиявок' ('Attack of the Giant Leeches'). Ничего удивительного!

In [8]:
%%timeit -r1 -n1
print(Rm.top_controversial(10))

{'Fanny and Alexander (Fanny och Alexander)': 10.12, "Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo)": 10.12, 'Lassie': 8.0, 'Zed & Two Noughts A': 8.0, 'Oscar': 8.0, 'Kwaidan (Kaidan)': 8.0, 'Troll 2': 6.75, 'Tenebre': 6.12, 'Hard Way The': 6.12, 'Attack of the Giant Leeches': 6.12}
81.2 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### *(БОНУС)* Распределение по рейтингу жанров. 
В лидерах здесь жанр нуар - кто не любит поэтичную и мрачную атмосферу нуарного кино? Следом за ним расположились военные фильмы и, с незначительным отставанием, документальное кино. В общем, всё, что любит смотреть твой отец. 

In [9]:
%%timeit -r1 -n1
print(Rm.average_ratings_by_genres())

{'Film-Noir': 3.92, 'War': 3.81, 'Documentary': 3.8, 'Drama': 3.66, 'Crime': 3.66, 'Animation': 3.63, 'Mystery': 3.63, 'IMAX': 3.62, 'Western': 3.58, 'Musical': 3.56, 'Adventure': 3.51, 'Romance': 3.51, 'Fantasy': 3.49, 'Thriller': 3.49, '(no genres listed)': 3.49, 'Sci-Fi': 3.46, 'Action': 3.45, 'Children': 3.41, 'Comedy': 3.38, 'Horror': 3.26}
126 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Распределение пользователей по количеству выставленных оценок.
Самым активным пользователем оказался юзер с id 414, он оценил аж 2698 фильмов.  По нашим подсчётам, если смотреть по 1 фильму в день, вам потребуется 7 лет и 5 месяцев. Настоящий киноманьяк! 

In [10]:
%%timeit -r1 -n1
print(Ru.number_of_ratings())

{414: 2698, 599: 2478, 474: 2108, 448: 1864, 274: 1346, 610: 1302, 68: 1260, 380: 1218, 606: 1115, 288: 1055, 249: 1046, 387: 1027, 182: 977, 307: 975, 603: 943, 298: 939, 177: 904, 318: 879, 232: 862, 480: 836, 608: 831, 600: 763, 483: 728, 590: 728, 105: 722, 19: 703, 305: 677, 489: 648, 111: 646, 438: 635, 217: 613, 140: 608, 477: 600, 555: 578, 91: 575, 28: 570, 219: 528, 534: 520, 89: 518, 64: 517, 226: 507, 561: 505, 18: 502, 525: 500, 57: 476, 381: 474, 368: 469, 509: 467, 469: 465, 560: 458, 462: 455, 292: 446, 21: 443, 597: 443, 42: 440, 160: 437, 294: 437, 580: 436, 596: 411, 202: 403, 275: 403, 517: 400, 45: 399, 156: 398, 514: 397, 391: 386, 567: 385, 357: 383, 103: 377, 339: 371, 62: 366, 199: 363, 125: 360, 51: 359, 132: 347, 66: 345, 313: 340, 200: 334, 221: 331, 6: 314, 453: 311, 50: 310, 425: 306, 428: 300, 573: 299, 352: 294, 84: 293, 122: 292, 382: 291, 356: 289, 135: 279, 239: 279, 365: 277, 484: 275, 104: 273, 63: 271, 325: 270, 169: 269, 290: 267, 332: 267, 495: 2

### Распределение пользователей на основе среднего значения их оценок.
Почётное место самого непритязательного кинолюбителя занял пользователь с id 53. Всем просмотренным фильмам он, не колеблясь, выставил 5.0.

In [11]:
%%timeit -r1 -n1
print(Ru.average_or_median_ratings())

{53: 5.0, 251: 4.87, 515: 4.85, 25: 4.81, 30: 4.74, 523: 4.69, 348: 4.67, 171: 4.63, 452: 4.56, 43: 4.55, 122: 4.55, 371: 4.55, 441: 4.52, 400: 4.51, 52: 4.48, 538: 4.47, 417: 4.46, 168: 4.46, 543: 4.45, 106: 4.44, 601: 4.43, 319: 4.43, 475: 4.41, 413: 4.41, 188: 4.4, 276: 4.39, 12: 4.39, 533: 4.38, 581: 4.38, 154: 4.38, 1: 4.37, 586: 4.37, 69: 4.37, 544: 4.36, 59: 4.36, 128: 4.36, 253: 4.35, 459: 4.35, 553: 4.34, 585: 4.34, 519: 4.33, 336: 4.32, 70: 4.32, 435: 4.32, 300: 4.3, 93: 4.29, 532: 4.28, 74: 4.27, 291: 4.26, 80: 4.26, 224: 4.26, 49: 4.26, 169: 4.25, 246: 4.25, 398: 4.25, 164: 4.25, 162: 4.24, 362: 4.24, 209: 4.24, 494: 4.23, 250: 4.22, 17: 4.21, 337: 4.21, 573: 4.21, 340: 4.21, 252: 4.2, 460: 4.2, 79: 4.2, 227: 4.2, 595: 4.2, 258: 4.2, 364: 4.19, 526: 4.19, 97: 4.19, 119: 4.18, 267: 4.18, 505: 4.18, 296: 4.17, 72: 4.16, 465: 4.16, 491: 4.16, 458: 4.15, 290: 4.14, 243: 4.14, 29: 4.14, 37: 4.14, 547: 4.14, 221: 4.13, 568: 4.13, 389: 4.12, 443: 4.12, 572: 4.12, 527: 4.12, 105: 4

### Топ 10 пользователей с наибольшим разбросом в их оценках.
У этих пользователей фильм вызывает либо любовь, либо ненависть, они не боятся поставить как и самую высокую оценку, так и заклеймить фильм низким баллом. 

In [12]:
%%timeit -r1 -n1
print(Ru.biggest_variance_ratings(10))

{3: 4.37, 55: 3.22, 461: 3.22, 259: 3.05, 329: 3.05, 175: 2.87, 502: 2.84, 598: 2.84, 393: 2.63, 138: 2.56}
119 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## Анализ данных Movies.csv

### Данный рейтинг отражает количество выпущенных фильмов по годам.
Самым плодотворным годом стал 2002 год с 311 фильмамим. Прекрасный год в истории кино, подаривший нам Властелина колец, Гарри Поттера и Людей в Чёрном.


In [13]:
%%timeit -r1 -n1
print(M.dist_by_release())

OrderedDict({'2002': 311, '2006': 295, '2001': 294, '2007': 284, '2000': 283, '2009': 282, '2003': 279, '2004': 279, '2014': 278, '1996': 276, '2015': 274, '2005': 273, '2008': 269, '1999': 263, '1997': 260, '1995': 259, '1998': 258, '2011': 254, '2010': 247, '2013': 239, '1994': 237, '2012': 233, '2016': 218, '1993': 198, '1992': 167, '1988': 165, '1987': 153, '1990': 147, '1991': 147, '2017': 147, '1989': 142, '1986': 139, '1985': 126, '1984': 101, '1981': 92, '1980': 89, '1982': 87, '1983': 83, '1979': 69, '1977': 63, '1973': 59, '1978': 59, '1965': 47, '1971': 47, '1974': 45, '1976': 44, '1964': 43, '1967': 42, '1968': 42, '1975': 42, '1966': 42, '2018': 41, '1962': 40, '1972': 39, '1963': 39, '1959': 37, '1960': 37, '1955': 36, '1969': 35, '1961': 34, '1970': 33, '1957': 33, '1958': 31, '1953': 30, '1956': 30, '1940': 25, '1949': 25, '1954': 23, '1942': 23, '1939': 23, '1946': 23, '1951': 22, '1950': 21, '1947': 20, '1948': 20, '1941': 18, '1936': 18, '1945': 17, '1937': 16, '1952

### Теперь подсчитаем количество фильмов, снятых в определённых жанрах.
Как думаете, какой жанр самый популярный? Конечно же драма! 4361 фильм! Все мы любим немного попереживать. Затем следует комедия, ведь смеяться мы тоже любим. А потом триллер, что бы уж точно испытать весь спектр эмоций. 

In [14]:
%%timeit -r1 -n1
print(M.dist_by_genres())

{'Drama': 4361, 'Comedy': 3756, 'Thriller': 1894, 'Action': 1828, 'Romance': 1596, 'Adventure': 1263, 'Crime': 1199, 'Sci-Fi': 980, 'Horror': 978, 'Fantasy': 779, 'Children': 664, 'Animation': 611, 'Mystery': 573, 'Documentary': 440, 'War': 382, 'Musical': 334, 'Western': 167, 'IMAX': 158, 'Film-Noir': 87, '(no genres listed)': 34}
18.8 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### *(BONUS)* Данный рейтинг отражает долю каждого жанра среди всех жанров на площадке.
Здесь видно, что если взять все фильмы жанров драма, комедия и триллер, то они будут составлять аж 45.34% от всех снятых кинолент! Почти половина!

In [15]:
%%timeit -r1 -n1
print(M.stat_by_genres()) 

{'Drama': '19.75 %', 'Comedy': '17.01 %', 'Thriller': '8.58 %', 'Action': '8.28 %', 'Romance': '7.23 %', 'Adventure': '5.72 %', 'Crime': '5.43 %', 'Sci-Fi': '4.44 %', 'Horror': '4.43 %', 'Fantasy': '3.53 %', 'Children': '3.01 %', 'Animation': '2.77 %', 'Mystery': '2.59 %', 'Documentary': '1.99 %', 'War': '1.73 %', 'Musical': '1.51 %', 'Western': '0.76 %', 'IMAX': '0.72 %', 'Film-Noir': '0.39 %', '(no genres listed)': '0.15 %'}
307 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 картин с максимально охваченным количеством жанров.
Первое место занял фильм под названием 'Шина' ('Rubber'), посягнувший аж на 10 разных жанров. Вы только посмотрите на его описание: "Неожиданно ожившая автомобильная шина начинает бесцельно болтаться по округе. Резина вдруг обнаруживает у себя неизвестно откуда взявшуюся способность убивать живые объекты на расстоянии. А увидев однажды гору горящих товарищей, покрышка решает направить свою агрессию на месть человечеству." С ума сойти, такое действительно сложно вписать в один жанр. 

In [16]:
%%timeit -r1 -n1
print(M.most_genres(10))

{'Rubber': 10, 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie)': 8, 'Mulan': 7, 'Who Framed Roger Rabbit?': 7, 'Osmosis Jones': 7, 'Interstate 60': 7, 'Robots': 7, 'Pulse': 7, 'Aqua Teen Hunger Force Colon Movie Film for Theaters': 7, 'Enchanted': 7}
30.6 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## Анализ тегов. Теги помогают понять, как пользователи описывают фильмы своими словами.

### Топ 10 тегов с наибольшим количеством слов. Ищем теги где много слов (иногда сложно описать свои впечатления одним словом)
- Самым длинным тегом является: "Something for everyone in this one... saw it without and plan on seeing it with kids". Пользователь посмотрел мультфильм "В гости к Робинсонам" и посчитал что каждый может найти в нём что-то для себя.
- Вторым по длине является "the catholic church is the most corrupt organization in history". Такая мысль возникла у пользователя после просмотра фильма "Избавь нас от лукавого". Тут, пожалуй, без комментариев.
- Третье место занял тег 'villain nonexistent or not needed for good story' - злодей отсутствует или не нужен для хорошей истории, так посчитал юзер после просмотра фильма о поросёнке Бейбе. И в чём он не прав?

In [17]:
%%timeit -r1 -n1
print(T.most_words(10))

{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16, 'the catholic church is the most corrupt organization in history': 10, 'villain nonexistent or not needed for good story': 8, '06 Oscar Nominated Best Movie - Animation': 7, 'It was melodramatic and kind of dumb': 7, 'Oscar (Best Music - Original Score)': 6, 'stop using useless characters for filler': 6, 'Oscar (Best Effects - Visual Effects)': 6, 'Everything you want is here': 5, 'based on a true story': 5}
4.05 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 самых длинных тегов по количеству символов. Ищем теги где много букв.
Здесь в тройке лидеров те же теги, что и в предыдущем рейтинге. Появился любопытный тег - 'r:sustained strong stylized violence', который можно примерно перевести как "длительные сцены жестокости в стилизованной форме". Страшно, очень страшно! Мы не знаем что это такое, если бы мы знали, что это такое, мы не знаем что это такое. После небольшого рисёча мы выяснили, что этот тег выставлен фильму Криминальное чтиво. Это всё объясняет.

In [18]:
%%timeit -r1 -n1
print(T.longest(10))

['Something for everyone in this one... saw it without and plan on seeing it with kids!', 'the catholic church is the most corrupt organization in history', 'villain nonexistent or not needed for good story', 'r:disturbing violent content including rape', '06 Oscar Nominated Best Movie - Animation', 'stop using useless characters for filler', 'Academy award (Best Supporting Actress)', 'Oscar (Best Effects - Visual Effects)', 'audience intelligence underestimated', 'It was melodramatic and kind of dumb']
2.08 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 пересечений между тегами с наибольшим количеством слов внутри и тегами с наибольшей длиной по количеству символов.
- первый в рейтинге тег 'Oscar Nominated Best Movie - Animation' - лучший анимационный фильм, номинированный на Оскар;
- второе место у уже знакомого нам тега 'villain nonexistent or not needed for good story';
- замыкает тройку лидеров тег 'It was melodramatic and kind of dumb'. "это было мелодраматично и немного глупо" так описан фильм "Выкуп" с Мелом Гибсоном. 

In [19]:
%%timeit -r1 -n1
print(T.most_words_and_longest(10))

['Something for everyone in this one... saw it without and plan on seeing it with kids!', 'villain nonexistent or not needed for good story', '06 Oscar Nominated Best Movie - Animation', 'stop using useless characters for filler', 'It was melodramatic and kind of dumb', 'the catholic church is the most corrupt organization in history', 'Oscar (Best Effects - Visual Effects)']
5.64 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Топ 10 самых популярных тегов. 
- самый популярный тег 'In Netflix queue' - 131 появление. "В моей подборке Netflix" или "В моей/вашей очереди просмотра Netflix";
- второй по частоте тег 'atmospheric' (атмосферный) - был выставлен 36 раз;
- третье место разделяют сразу два тега 'superhero' (супергерой) и 'thought-provoking' (наводящий на размышления) - по 24 появления у каждого тега.

In [20]:
%%timeit -r1 -n1
print(T.most_popular(10))

{'In Netflix queue': 131, 'atmospheric': 36, 'superhero': 24, 'thought-provoking': 24, 'funny': 23, 'Disney': 23, 'surreal': 23, 'religion': 22, 'sci-fi': 21, 'dark comedy': 21}
2.69 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Поиск уникальных тегов по слову, переданному в качестве аргумента.
Мы выбрали слово 'dark' и нашли аж 5 тегов, его содержащих.
Как думаете, найдутся ли теги со словом 'Python'? Как ни странно - да! Всего один тег - 'Monty Python'. Ведь создатель языка Python Гвидо ван Россум назвал его в честь популярного британского комедийного шоу «Летающий цирк Монти Пайтона». 

In [21]:
%%timeit -r1 -n1
print(T.tags_with("dark"))

['dark', 'dark comedy', 'dark fairy tale', 'dark hero', 'dark humor']
1.28 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Метод, который возвращает список тегов для указанного фильма по его id
Вот теги для фильма с id 2 ("Джуманджи") - 'fantasy', 'robin williams', 'game', 'magic board game'. Легендарный фильм, все помнят этот пугающий тревожный звук барабанов. 

In [22]:
%%timeit -r1 -n1
print(T.get_tags_by_the_movie(2))

['robin williams', 'fantasy', 'game', 'magic board game']
1.22 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## Анализ данных о фильмах по информации с сайта IMDB

### Парсинг информации о фильме с сайта [IMDB](https://www.imdb.com)
Этот метод извлекает информацию с сайта для указанного списка id фильмов. 

In [23]:
%%timeit -r1 -n1
print(L.get_imdb([92, 22, 32, 42, 58, 69], 
        ["Title", "Runtime", "Director","Budget", 
        "Cumulative worldwide gross"]))

[[92, 'Mary Reilly', '1 hour 48 minutes', 'Stephen Frears', '$47,000,000', '$12,379,402'], [69, 'Friday', '1 hour 31 minutes', 'F. Gary Gray', '$3,500,000', '$28,215,918'], [58, 'The Postman', '1 hour 48 minutes', 'Michael Radford', '$3,000,000', '$21,901,256'], [42, 'Dead Presidents', '1 hour 59 minutes', 'Albert Hughes, Allen Hughes', '$10,000,000', '$24,147,179'], [32, '12 Monkeys', '2 hours 9 minutes', 'Terry Gilliam', '$29,000,000', '$168,839,459'], [22, 'Copycat', '2 hours 3 minutes', 'Jon Amiel', '$20,000,000', '$32,051,917']]
8.66 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Ищем режиссёров с самыми большими фильмографиями
#### Парсим n страниц с количеством созданных картин среди режиссеров, сортируем по убыванию количества фильмов. 
Чтобы корректно отобразить рейтинг режиссёров по их фильмографиям понадобится извлечь данные для всех фильмов из списка links.csv, а их 9742! На это уйдёт пара часов. Так что мы ограничили выборку сотней строк. И в нашем скромном списке из 100 картин отличился Мартин Скорсезе.

In [25]:
%%timeit -r1 -n1
print(L.top_directors(10))

{'Martin Scorsese': 2, 'Stanley Tong': 1, 'Mel Gibson': 1, 'Robert Leacock': 1, 'Brian Henson': 1, 'Doris Dörrie': 1, 'Clint Eastwood': 1, 'Dennis Dugan': 1, 'John Dahl': 1, 'Nick Castle': 1}
1min 50s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


### Ищем самые высокобюджетные фильмы
#### Парсим n картин и отражаем информацию по затраченному бюджету на их создание, сортируем по убыванию бюджета. 
На первом месте здесь "Остров головорезов" ("Cutthroat Island") с бюджетом $98,000,000. Масштабная сцена сражения между двумя пиратскими кораблями была снята на большом искусственном озере на Мальте, никакого CGI. Нетрудно догадаться, что бюджет такого фильма впечатляющий. 

In [26]:
%%timeit -r1 -n1
print(L.most_expensive(100))

{'Cutthroat Island': '$98,000,000', 'Braveheart': '$72,000,000', 'Money Train': '$68,000,000', 'Jumanji': '$65,000,000', 'American President, The': '$62,000,000', 'GoldenEye': '$60,000,000', 'Heat': '$60,000,000', 'Sabrina': '$58,000,000', 'Pocahontas': '$55,000,000', 'Casino': '$52,000,000', 'Broken Arrow': '$50,000,000', 'Fair Game': '$50,000,000', 'Assassins': '$50,000,000', 'Mary Reilly': '$47,000,000', 'Indian in the Cupboard, The': '$45,000,000', 'Juror, The': '$44,000,000', 'Nixon': '$44,000,000', 'City Hall': '$40,000,000', 'White Squall': '$38,000,000', 'Sudden Death': '$35,000,000', 'Nick of Time': '$33,000,000', 'Seven (a.k.a. Se7en)': '$33,000,000', 'Get Shorty': '$30,250,000', 'Babe': '$30,000,000', 'Ace Ventura: When Nature Calls': '$30,000,000', 'Dracula: Dead and Loving It': '$30,000,000', 'Father of the Bride Part II': '$30,000,000', 'Toy Story': '$30,000,000', 'Twelve Monkeys (a.k.a. 12 Monkeys)': '$29,000,000', 'Grumpier Old Men': '$25,000,000', 'Bridges of Madison C

### Ищем самые кассовые фильмы
#### Парсим n картин и отражаем разницу между совокупным мировым кассовым сбором и бюджетом.
Из представленной сотни кинолент самой прибыльной оказалась "История игрушек", она принесла создателям $364,436,586

In [60]:
%%timeit -r1 -n1
print(L.most_profitable(100))

{'Toy Story': '$364,436,586', 'Seven (a.k.a. Se7en)': '$295,981,827', 'GoldenEye': '$292,194,034', 'Pocahontas': '$291,079,773', 'Babe': '$224,134,910', 'Jumanji': '$197,821,940', 'Ace Ventura: When Nature Calls': '$182,385,533', 'Bridges of Madison County, The': '$158,016,617', 'Dangerous Minds': '$156,519,401', 'Braveheart': '$141,216,216', 'Twelve Monkeys (a.k.a. 12 Monkeys)': '$139,839,459', 'Heat': '$127,436,818', 'Sense and Sensibility': '$118,582,776', 'Mortal Kombat': '$104,195,920', 'Broken Arrow': '$100,270,147', 'Get Shorty': '$84,851,622', "Mr. Holland's Opus": '$83,269,971', 'Waiting to Exhale': '$65,452,156', 'Casino': '$64,112,375', 'Father of the Bride Part II': '$46,594,107', 'Grumpier Old Men': '$46,518,503', 'American President, The': '$45,879,496', 'Clueless': '$44,690,454', 'How to Make an American Quilt': '$31,200,020', 'Happy Gilmore': '$29,422,354', 'Sudden Death': '$29,350,171', 'Leaving Las Vegas': '$28,429,928', 'Dead Man Walking': '$28,363,635', 'Taxi Driver

### Ищем самые продолжительные фильмы
#### Парсим n картин и отражаем информацию по самым продолжительным.
В представленной выборке самым длинным оказался фильм 'Никсон' ('Nixon') - биографический фильм о бывшем президенте США. Он длится аж 3 часа 12 минут. Фильм для самыйх стойких кинолюбителей. Ну, или для фанатов Никсона.   

In [61]:
%%timeit -r1 -n1
print(L.longest(100))

{'Nixon': '3 hours 12 minutes', 'White Squall': '2 hours 9 minutes', 'Twelve Monkeys (a.k.a. 12 Monkeys)': '2 hours 9 minutes', 'Seven (a.k.a. Se7en)': '2 hours 7 minutes', 'Sabrina': '2 hours 7 minutes', 'Braveheart': '2 hours 58 minutes', 'Casino': '2 hours 58 minutes', 'Misérables, Les': '2 hours 55 minutes', 'Heat': '2 hours 50 minutes', 'Cutthroat Island': '2 hours 4 minutes', 'Waiting to Exhale': '2 hours 4 minutes', 'Othello': '2 hours 3 minutes', 'Copycat': '2 hours 3 minutes', "Mr. Holland's Opus": '2 hours 24 minutes', 'Dead Man Walking': '2 hours 2 minutes', 'Sense and Sensibility': '2 hours 16 minutes', 'Bridges of Madison County, The': '2 hours 15 minutes', 'Assassins': '2 hours 13 minutes', 'GoldenEye': '2 hours 10 minutes', 'Nico Icon': '1 hour 7 minutes', 'Dead Presidents': '1 hour 59 minutes', 'Juror, The': '1 hour 58 minutes', 'Unforgettable': '1 hour 57 minutes', 'How to Make an American Quilt': '1 hour 57 minutes', 'Restoration': '1 hour 57 minutes', 'Angels and Ins

### Ищем фильмы с самой высокой поминутной стоимостью. Время - деньги!
#### Парсим n картин и делим их бюджет на их продолжительность по времени.
В данной выборке и 100 фильмов первое место снова занял "Остров головорезов". На втором месте "Покахонтас" - Дисней не пожалел денег на качественную анимацию. Третье место у "Jumanji" - каждая минута фильма обходилась создателям в копеечку. 

In [62]:
%%timeit -r1 -n1
print(L.top_cost_per_minute(100))

{'Cutthroat Island': 790322.58, 'Pocahontas': 679012.35, 'Jumanji': 625000.0, 'Money Train': 618181.82, 'Fair Game': 549450.55, 'American President, The': 543859.65, 'Indian in the Cupboard, The': 468750.0, 'Broken Arrow': 462962.96, 'GoldenEye': 461538.46, 'Sabrina': 456692.91, 'Mary Reilly': 435185.19, 'Braveheart': 404494.38, 'Assassins': 375939.85, 'Juror, The': 372881.36, 'Nick of Time': 370786.52, 'Toy Story': 370370.37, 'City Hall': 360360.36, 'Heat': 352941.18, 'Dracula: Dead and Loving It': 340909.09, 'Ace Ventura: When Nature Calls': 333333.33, 'Babe': 329670.33, 'Sudden Death': 315315.32, 'White Squall': 294573.64, 'Casino': 292134.83, 'Get Shorty': 288095.24, 'Father of the Bride Part II': 283018.87, 'Seven (a.k.a. Se7en)': 259842.52, 'Grumpier Old Men': 247524.75, 'Dangerous Minds': 232323.23, 'Nixon': 229166.67, 'Twelve Monkeys (a.k.a. 12 Monkeys)': 224806.2, 'Vampire in Brooklyn': 200000.0, 'Eye for an Eye': 198019.8, 'Home for the Holidays': 194174.76, 'To Die For': 188

### Режиссёры, которые не любят экономить. 
#### *(BONUS)* Парсим n картин и отражаем информацию по режиссерам, которые сняли самые высокобюджетные фильмы.
- первое место в данной подборке занимает режиссёр Рэнни Харлин. Вы слышали о таком? Мы тоже нет. А он как раз таки и снял "Остров головорезов".
- второе место занял Мел Гибсон. Мы все знаем его как актёра, а он, оказывается, не тлько снимается, но и сам снимает кино.
- третье место у Джозефа Рубина. 

In [63]:
%%timeit -r1 -n1
print(L.directors_of_most_expensive(100))

{'Renny Harlin': '$98,000,000', 'Mel Gibson': '$72,000,000', 'Joseph Ruben': '$68,000,000', 'Joe Johnston': '$65,000,000', 'Rob Reiner': '$62,000,000', 'Martin Campbell': '$60,000,000', 'Michael Mann': '$60,000,000', 'Sydney Pollack': '$58,000,000', 'Mike Gabriel': '$55,000,000', 'Martin Scorsese': '$52,000,000', 'John Woo': '$50,000,000', 'Andrew Sipes': '$50,000,000', 'Richard Donner': '$50,000,000', 'Stephen Frears': '$47,000,000', 'Frank Oz': '$45,000,000', 'Brian Gibson': '$44,000,000', 'Oliver Stone': '$44,000,000', 'Harold Becker': '$40,000,000', 'Ridley Scott': '$38,000,000', 'Peter Hyams': '$35,000,000', 'John Badham': '$33,000,000', 'David Fincher': '$33,000,000', 'Barry Sonnenfeld': '$30,250,000', 'Chris Noonan': '$30,000,000', 'Steve Oedekerk': '$30,000,000', 'Mel Brooks': '$30,000,000', 'Charles Shyer': '$30,000,000', 'John Lasseter': '$30,000,000', 'Terry Gilliam': '$29,000,000', 'Howard Deutch': '$25,000,000', 'Clint Eastwood': '$24,000,000', 'Stephen Herek': '$23,000,00